### PR TITLE
Add blas compare example

### PR DESCRIPTION
**Pull request by robieta**
Adds a standalone script which can be used to test different BLAS libraries. Right now I've deliberately kept it limited (only a couple BLAS libs and only GEMM and GEMV). It's easy enough to expand later.

CC @ngimel

### CODE REVIEW

File: torch/utils/benchmark/examples/blas_compare_setup.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        check_run_stdout = check_run.stdout.decode('utf-8')
+        print(check_run_stdout)
+
+        for e in env_spec.environment_variables:
+            if "BLAS" in e:
+                assert e in check_run_stdout, f"PyTorch build did not respect `BLAS=...`: {e}"
+
+        for s in env_spec.expected_blas_symbols:
+            assert s in check_run_stdout
+
```
Comment (ngimel) [inline torch/utils/benchmark/examples/blas_compare_setup.py:207]: do you think it also makes sense to check mkl version? `torch.__config__` provides it.

File: torch/utils/benchmark/examples/blas_compare.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+        shapes = (
+            # Square MatMul
+            ((n, n), (n, n), "(n x n) x (n x n)", "GEMM"),
+
+            # Matrix-Vector product
+            ((n, n), (n, 1), "(n x n) x (n x 1)", "GEMV"),
+        )
+        for (dtype_prefix, dtype), (x_shape, y_shape, shape_str, blas_type) in it.product(dtypes, shapes):
+            t = Timer(
+                stmt="torch.mm(x, y)",
```
[the remark this answers is not in the document]
Comment (robieta) [reply torch/utils/benchmark/examples/blas_compare.py:84]: Hmm. I changed the names to be more generic. I do see qualitative differences: OpenBLAS is competitive on MMP but lags significantly with MVP, and callgrind confirms that MKL is smart enough to switch to gemv under the hood.

File: torch/utils/benchmark/examples/blas_compare.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        shell=True
+    )
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--DETAIL_in_subprocess", action="store_true")
+    parser.add_argument("--DETAIL_in_compare", action="store_true")
+    parser.add_argument("--DETAIL_seed", type=int, default=None)
+    parser.add_argument("--DETAIL_num_threads", type=int, default=None)
```
[the remark this answers is not in the document]
Comment (robieta) [reply torch/utils/benchmark/examples/blas_compare.py:210]: I added a comment clarifying that these are used to direct subprocesses; they're not meant for users. (Hence the `DETAIL_` prefix.)

File: torch/utils/benchmark/examples/blas_compare_setup.py (hunk 1 above)
Comment (robieta) [reply torch/utils/benchmark/examples/blas_compare_setup.py:213]: Done.

**Comment by robieta** [comment]
Sorry for the delay in addressing your comments.

**Review by ngimel — APPROVED** [review]

**Comment by robieta** [comment]
I'm surprised to see that MyPy is checking this at all.

**Pull request closed without merging**